components/applications-service/pkg/generator/runner.go

```diff
@@ -0,0 +1,253 @@
+		UUID:              uuid,
+		Cfg:               cfg,
+		MessagePrototypes: s.MessagePrototypes,
+		Stats:             stats,
+	}
+
+	err = sup.Run()
+	if err != nil {
+		fmt.Printf("Supervisor failed to run: %s\n", err)
+		return err
+	}
+
+	return nil
+}
+
+type SupSim struct {
+	Name              string
+	Idx               int32
+	Cfg               *RunnerConfig
+	UUID              uuid.UUID
+	MessagePrototypes []*MessagePrototype
+	Stats             *RunnerStatsKeeper
+	nc                *nats.NatsClient
+}
+
+func (s *SupSim) Run() error {
+	s.Stats.SupStarted()
+	defer s.Stats.SupDied()
+
+	url := fmt.Sprintf("nats://%s@%s:4222", s.Cfg.AuthToken, s.Cfg.Host)
+	cluster := "event-service"
+	client := fmt.Sprintf("load-generator-%s", s.UUID)
+	durable := client
+	subject := "habitat"
+
+	s.nc = nats.NewExternalClient(url, cluster, client, durable, subject)
+	s.nc.InsecureSkipVerify = true
+	err := s.nc.Connect()
+	if err != nil {
+		return err
+	}
+	defer s.nc.Close()
+	// run loadgen loop
+
+	log.WithFields(log.Fields{"name": s.Name, "index": s.Idx, "uuid": s.UUID.String()}).Info("starting supervisor")
+
+	// the ticker always waits its full time before publishing. Also we wish to
+	// splay out the simulated sups for a more even distribution of messages. So
+	// we get a random splay, publish once, then loop.
+	splay := rand.Int31n(int32(s.Cfg.Tick))
+	time.Sleep(time.Duration(splay) * time.Second)
+	s.PublishAll()
+
+	ticker := time.NewTicker(time.Duration(s.Cfg.Tick) * time.Second)
+	for _ = range ticker.C {
+		s.PublishAll()
+	}
+	return nil
+}
+
+func (s *SupSim) PublishAll() error {
+	log.WithFields(log.Fields{
+		"name":          s.Name,
+		"index":         s.Idx,
+		"uuid":          s.UUID.String(),
+		"service_count": len(s.MessagePrototypes),
+	}).Info("publishing messages")
+
+	for _, m := range s.MessagePrototypes {
+
+		msg := m.CreateMessage(s.UUID.String())
+
+		log.WithFields(log.Fields{
+			"name":          s.Name,
+			"index":         s.Idx,
+			"uuid":          s.UUID.String(),
+			"service_count": len(s.MessagePrototypes),
+			"pkg_fqid":      fmt.Sprintf("%s/%s/%s/%s", m.Origin, m.PkgName, versionNumber, m.Release),
+		}).Debug("publishing messages")
+
+		err := s.nc.PublishHabService(msg)
+		if err != nil {
+			s.Stats.FailedPublish()
+			return err
+		}
+		s.Stats.SuccessfulPublish()
+	}
+	return nil
+}
+
+// Data elements of a message that currently exist:
+// sup-id         AUTO
+// group          hardcode to "default"
+// application    assign per-service in sup template
+// environment    <name>   The environment name of the current deployment
+// health         hardcode to 0 (OK) for now.
+// status         hardcode to 0 (RUNNING) for now
+//
+// Package Indentifier
+// --origin      parse from service name
+// --name        parse from service name
+// --version     hardcode to 1.0.0 for now
+// --release     generate from the startup time of the generator
+type MessagePrototype struct {
```

Comment:
Why are we using this message instead of the `HabService` proto message? (curious)

Reply:
the `MessagePrototype` will always have only a subset of the fields that the `HabService` message has, and I wanted to be more explicit about how the `MessagePrototype` gets converted into individual `HabService` messages rather than detecting golang zero values and treating those as special